File: 235. Lowest Common Ancestor of a Binary Search Tree.swift
//  Definition for a binary tree node.
public class TreeNode {
    public var val: Int
    public var left: TreeNode?
    public var right: TreeNode?
    public init() { self.val = 0; self.left = nil; self.right = nil; }
    public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
    public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
      self.val = val
      self.left = left
      self.right = right
  }
}

 
class Solution {
    
    func lowestCommonAncestor(_ root: TreeNode?, _ p: TreeNode?, _ q: TreeNode?) -> TreeNode? {
        // Should never execute
        guard let root = root else {
            return nil
        }
        if p!.val > root.val && q!.val > root.val {
            return lowestCommonAncestor(root.right, p, q)
        }else if p!.val < root.val && q!.val < root.val {
            return lowestCommonAncestor(root.left, p, q)
        }else{
            return root
        }
    }
    
}

File: 235. Lowest Common Ancestor of a Binary Search Tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:

    # Time complexity: O(logN) because we use recursion according to the BST property
    def lowestCommonAncestorLogNSolution(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root or root.val == p.val or root.val == q.val: return root
        if root.val < p.val and root.val < q.val:  return self.lowestCommonAncestor(root.right,p,q)
        elif root.val > p.val and root.val > q.val: return self.lowestCommonAncestor(root.left,p,q)
        else: return root

    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        
        if root == None: return None
        if root.val == p.val or root.val == q.val: return root
        left = self.lowestCommonAncestor(root.left,p,q)
        right = self.lowestCommonAncestor(root.right,p,q)
        if left and right: return root
        return left if left != None else right
